grpc4bmi/bmi_client_apptainer.py

```diff
@@ -0,0 +1,273 @@
+
+    On instantization launches an Apptainer container.
+    The Apptainer container image is expected to run a BMI GRPC server as its default command.
+    The client picks a random port and expects the container to run the BMI GRPC server on that port.
+    The port is passed to the container using the BMI_PORT environment variable.
+
+    Args:
+        image: Apptainer image.
+
+            For Docker Hub image use `docker://*` or convert it to a Apptainer image file.
+
+            To convert Docker image
+            `ewatercycle/walrus-grpc4bmi <https://hub.docker.com/layers/ewatercycle/walrus-grpc4bmi>`_
+            with `v0.2.0` tag to `./ewatercycle-walrus-grpc4bmi_v0.2.0.sif` Apptainer image file use:
+
+            .. code-block:: console
+
+              apptainer pull ewatercycle-walrus-grpc4bmi_v0.2.0.sif docker://ewatercycle/walrus-grpc4bmi:v0.2.0
+
+        input_dirs (Iterable[str]): Directories for input files of model.
+
+            All of directories will be mounted read-only inside Apptainer container on same path as outside container.
+
+        work_dir (str): Working directory for model.
+
+            Directory is mounted inside container and changed into.
+
+            To create a random work directory you could use
+
+            .. code-block:: python
+
+              from tempfile import TemporaryDirectory
+              from grpc4bmi.bmi_client_apptainer import BmiClientApptainer
+
+              work_dir = TemporaryDirectory()
+
+              image = 'ewatercycle-walrus-grpc4bmi_v0.2.0.sif'
+              client =  BmiClientApptainer(image, work_dir.name)
+
+              # Write config to work_dir and interact with client
+
+              # After checking output in work_dir, clean up
+              work_dir.cleanup()
+
+        delay (int): Seconds to wait for Apptainer container to startup, before connecting to it
+
+            Increase when container takes a long time to startup.
+
+        timeout (int): Seconds to wait for gRPC client to connect to server.
+
+            By default will try forever to connect to gRPC server inside container.
+            Set to low number to escape endless wait.
+
+        capture_logs (bool): Whether to capture stdout and stderr of container .
+
+            If false then redirects output to null device never to be seen again.
+            If true then redirects output to temporary file which can be read with :py:func:`BmiClientApptainer.logs()`.
+            The temporary file gets removed when this object is deleted.
+
+    **Example 1: Config file already inside image**
+
+    MARRMoT has an `example config file <https://github.com/wknoben/MARRMoT/blob/master/BMI/Config/BMI_testcase_m01_BuffaloRiver_TN_USA.mat>`_ inside its Docker image.
+
+    .. code-block:: python
+
+        from grpc4bmi.bmi_client_apptainer import BmiClientApptainer
+        client = BmiClientApptainer(image='docker://ewatercycle/marrmot-grpc4bmi:latest')
```

Comment:
This line won't run. work_dir is a required argument;
`TypeError: __init__() missing 1 required positional argument: 'work_dir'`

Reply:
The next 3 lines run fine if a work_dir is given.